cdap-authorization-ldap-role/src/main/java/io/cdap/cdap/security/authorization/ldap/role/RoleAuthorization.java

```diff
@@ -0,0 +1,292 @@
+import io.cdap.cdap.security.spi.authorization.UnauthorizedException;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.Properties;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of {@link AccessController} to use LDAP and role mapping to provide RBAC
+ */
+public class RoleAuthorization implements AccessController {
+  private static final Logger LOG = LoggerFactory.getLogger(RoleAuthorization.class);
+
+  // Is necessary for scheduled tasks
+  private static final String EMPTY_USER = "";
+
+  private LDAPSearcher searcherService;
+  private GroupWithRolesProvider roleProvider;
+
+  private boolean ignoreSystemUser;
+  private boolean loggingOnly;
+  private boolean disablePermissionsPropagation;
+
+  private String systemUsername;
+
+  public RoleAuthorization() {
+    systemUsername = RoleAuthorizationConstants.DEFAULT_SYSTEM_USER;
+  }
+
+  @Override
+  public void initialize(AuthorizationContext context) {
+    Properties properties = context.getExtensionProperties();
+    ignoreSystemUser = RoleAuthorizationUtil.getIgnoreSystemUserValue(properties);
+    loggingOnly = RoleAuthorizationUtil.getLoggingOnlyValue(properties);
+    disablePermissionsPropagation = RoleAuthorizationUtil.getDisablePermissionsPropagationValue(properties);
+
+    LOG.info("Permission propagation is: '{}'", !disablePermissionsPropagation);
+
+    setSystemUser();
+
+    if (loggingOnly) {
+      return;
+    }
+
+    searcherService = createLDAPSearcher(context);
+    searcherService.testConnection();
+
+    roleProvider = RoleAuthorizationUtil.createLDAPGroupRoleProvider(properties);
+    roleProvider.start();
+  }
+
+  @Override
+  public void destroy() {
+    if (!loggingOnly) {
+      roleProvider.stop();
+    }
+  }
+
+  @Override
+  public void enforce(EntityId entity, Principal principal, Set<? extends Permission> permissions)
+    throws AccessException {
+    String username = principal.getName();
+    LOG.debug("enforce user: '{}' entity: '{}' permissions: '{}'", username, entity, permissions);
+
+    // If we are not system user or blank user for running scheduled pipelines
+    if ((username.equals(systemUsername) && !ignoreSystemUser) || loggingOnly || username.equals(EMPTY_USER)) {
+      return;
+    }
+
+    PrincipalPermissions principalPermissions = getAllPermissions(username);
+    Set<? extends Permission> userPermissionsList;
+
+    // Propagation permissions if entity is not set and propagation is not disabled
+    if (!disablePermissionsPropagation && roleProvider.isEntityUnSet(entity.getEntityType())) {
+      userPermissionsList = RoleAuthorizationUtil.getPropagatedPermissions(entity, permissions, principalPermissions);
```

Comment:
I don't think you have to generate a list of propagated permissions (error message will confuse). You just have to check for namespace permission